Source/Core/DolphinWX/ISOFile.h

```diff
@@ -36,78 +36,101 @@ class GameListItem
   void DoState(PointerWrap& p);
-  bool ReloadBannerIfNeeded();
+  bool BannerChanged();
+  void BannerCommit();
+  bool EmuStateChanged();
+  void EmuStateCommit();
 
 private:
+  struct EmuState
+  {
+    int rating{};
+    std::string issues{};
+    bool operator!=(const EmuState& rhs) const
+    {
+      return rating != rhs.rating || issues != rhs.issues;
+    }
+    void DoState(PointerWrap& p);
+  };
+  struct Banner
+  {
+    std::vector<u8> buffer{};
+    int width{};
+    int height{};
+    bool empty() const { return buffer.empty(); }
+    void DoState(PointerWrap& p);
+  };
+
   bool IsElfOrDol() const;
-  // Outputs to m_pImage
-  void ReadVolumeBanner(const std::vector<u32>& buffer, int width, int height);
-  // Outputs to m_image
+  void ReadVolumeBanner(std::vector<u8>* image, const std::vector<u32>& buffer, int width,
+                        int height);
+  // Outputs to m_banner_wx
   bool SetWxBannerFromPngFile(const std::string& path);
-  void SetWxBannerFromRaw();
+  void SetWxBannerFromRaw(const Banner& banner);
 
-  // IMPORTANT: All data members must be save/restored in DoState.
+  // IMPORTANT: Nearly all data members must be save/restored in DoState.
   // If anything is changed, make sure DoState handles it properly and
   // GameListCtrl::CACHE_REVISION is incremented.
 
-  std::string m_FileName;
-
-  std::map<DiscIO::Language, std::string> m_names;
-  std::map<DiscIO::Language, std::string> m_descriptions;
-  std::string m_company;
-
-  std::string m_game_id;
-  u64 m_title_id = 0;
-
-  std::string m_issues;
-  int m_emu_state;
-
-  u64 m_FileSize;
-  u64 m_VolumeSize;
-
-  DiscIO::Region m_region;
-  DiscIO::Country m_Country;
-  DiscIO::Platform m_Platform;
-  DiscIO::BlobType m_blob_type;
-  u16 m_Revision;
-
-  wxImage m_image;
-  bool m_Valid;
-  std::vector<u8> m_pImage;
-  int m_ImageWidth, m_ImageHeight;
-  u8 m_disc_number;
-
-  std::string m_custom_name_titles_txt;  // Custom title from titles.txt
-  std::string m_custom_name;             // Custom title from INI or titles.txt
-  bool m_has_custom_name;
+  bool m_valid{};
+  std::string m_file_name{};
+
+  u64 m_file_size{};
+  u64 m_volume_size{};
+
+  std::map<DiscIO::Language, std::string> m_names{};
+  std::map<DiscIO::Language, std::string> m_descriptions{};
+  std::string m_company{};
+  std::string m_game_id{};
+  u64 m_title_id{};
+
+  DiscIO::Region m_region{};
+  DiscIO::Country m_country{};
+  DiscIO::Platform m_platform{};
+  DiscIO::BlobType m_blob_type{};
+  u16 m_revision{};
+  u8 m_disc_number{};
+
+  Banner m_banner{};
+  EmuState m_emu_state{};
+  // Overridden name from TitleDatabase
+  std::string m_custom_name{};
+
+  // wxImage is not handled in DoState
+  wxImage m_banner_wx{};
+
+  // The following data members allow GameListCtrl to construct new GameListItems in a threadsafe
+  // way. They should not be handled in DoState.
+  struct
+  {
+    EmuState emu_state;
+    Banner banner;
+  } m_pending{};
```

Comment:
Oh, okay. It makes sense that it doesn't work then.

I would strongly prefer to keep the purge cache option (and fix it so that it also purges the in-memory cache). It's very useful for people who edit either titles.txt or the name/banner data inside the games themselves, and it's also useful for developers who are making changes to the game parsing code. People like that are admittedly a minority, but I think the option is worth keeping around regardless, because it shouldn't be too hard to implement correctly.